src/midi/psg.c

```diff
@@ -0,0 +1,224 @@
+                    REG_SOUND1CNT_L = sweep;
+                    *envReg = (volume << 12) | (instrument->dutyTone << 6) | (instrument->length);
+                    *freqReg = reset | frequency;
+                    dummy = 0;
+                    dummy = 0;
+                    dummy = 0;
+                    dummy = 0;
+                    dummy = 0;
+                    dummy = 0;
+                    dummy = 0;
+                    dummy = 0;
+                    *freqReg = frequency | SOUNDCNTX_SOUND_RESET;
+                } else if (instrument->sweep == 0) {
+                    *envReg = (volume << 12) | (instrument->dutyTone << 6) | (instrument->length);
+                    *freqReg = reset | frequency;
+                }
+                break;
+
+            case PSG_PULSE_CHANNEL_2:
+                *envReg = (volume << 12) | (instrument->dutyTone << 6) | (instrument->length);
+                *freqReg = reset | frequency;
+                break;
+
+            case PSG_WAVE_CHANNEL:
+                if (sChannelJustStarted[id]) {
+                    REG_SOUNDCNT_L = 0xFF77; // Enable Stereo for All PSG Channels
+                    frequency |= SOUNDCNTX_SOUND_RESET;
+                    if ((u32 *)sPrevWavetable != instrument->wavetable) {
+                        volatile u32 *waveDest = &REG_SGWR0;
+                        u32 *waveSrc = instrument->wavetable;
+                        u32 waveSelect, waveControls;
+
+                        *waveDest++ = *waveSrc++;
+                        *waveDest++ = *waveSrc++;
+                        *waveDest++ = *waveSrc++;
+                        *waveDest++ = *waveSrc++;
+
+                        waveSelect = REG_SOUND3CNT_L;
+                        waveControls = SOUNDCNT3_BANK_SELECT;
+                        waveControls &= ~waveSelect;
+                        waveControls |= SOUNDCNT3_OUTPUT_ENABLE;
+                        REG_SOUND3CNT_L = waveControls;
+                    }
+                }
+                *envReg = (midi_psg_wave_vol_table[volume >> 2]) | (instrument->length);
+                *freqReg = frequency;
+                break;
+
+            case PSG_NOISE_CHANNEL:
+                *envReg = (volume << 12) | (instrument->length);
+                *freqReg = reset | frequency;
+                break;
+        }
+
+        sChannelPrevVolReg[id] = volume;
+        sChannelPrevFreqReg[id] = frequency;
+    }
+
+    sChannelJustStarted[id] = FALSE;
+
+    if (midi_note_update_adsr(soundChannel)) {
+        sChannelPrevVolReg[id] = 0;
+        *envReg = 0;
+        sChannelPrevFreqReg[id] = SOUNDCNTX_SOUND_RESET;
+        *freqReg = SOUNDCNTX_SOUND_RESET;
+        soundChannel->active = FALSE;
+    }
+}
+
+
+// Update All PSG SoundChannels
+void midi_psg_update(void) {
+    struct SoundChannel *channel = gMidiPSGChannelPool;
+    u16 controls = 0;
+    u32 i;
+
+    for (i = 0; i < 4; i++) {
```

Comment:
```suggestion
    for (i = 0; i < ARRAY_COUNT(gMidiPSGChannelPool); i++) {
```
Altough I think adding a define for PSG channel amount would be better